AITER Native Padding Support and BSHD + Padding --> THD + Padding conversion

Micky774 opened the pull request:
# Description

Feature update PR which includes several iterative changes for client-driven optimization targets. This PR includes both API changes for CK/AITER as well as changes in internal integration. See the list of changes for specifics.

Note that this will not be ready for merger until https://github.com/ROCm/aiter/pull/1212 is merged in and this PR's AITER commit is updated.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Integrated support for native padding kernels in fwd/bwd
- Added BSHD + Padding --> THD + Padding conversion mechanism
- Streamlined memory allocation logic
- Added runtime `max_seqlen` calculation gated by new env var `NVTE_CK_RUNTIME_MAX_SEQLEN`
- Adds `v3_api_check` support (temporary)
- Implements new AITER/CK API
- Update MQA post-processing kernels
- Remove `pad_between_seqs` (need to follow-up with a PR cleaning up test suite for old `pad_between_seqs` edge-cases)
- Added `NVTE_CK_RUNTIME_NUM_SEGMENTS` to guard runtime-calculation of the number of segments in the JAX integration

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

wangye805 committed:
[ROCm] manually pick up fwd native padding support from Meekail's PR

Micky774 committed:
Initial update

Micky774 committed:
Updated stride

Micky774 committed:
Corrected typing in allocation portions

Micky774 committed:
Applied Ye's patch

wangye805 committed:
[ROCm] manually pick Meekail's PR to support native padding for bwd

wangye805 committed:
[ROCm] manually pick Meekail's PR to support native padding for bwd

[ROCm] support v2 bwd native padding

wangye805 committed:
[ROCm] jax use runtime segment

wangye805 committed:
[ROCm] get runtime max_seqlen as well

wangye805 committed:
[ROCm] support v2 bwd native padding

Micky774 committed:
Updated conversion to include bwd pass

Micky774 committed:
Merge branch 'yewang12/te_aiter_native_padding_bwd' into zain/aiter-bwd-bshd-thd

Micky774 committed:
Added BWD BSHD-->THD conversion and minor logic refactor

Micky774 committed:
Corrected softmax lse bug

Micky774 committed:
Updated logic flow and re-caclulation

Micky774 committed:
Merge branch 'zain/aiter-bwd-bshd-thd' into zain/aiter-native-bshd-thd

Micky774 committed:
Added env var guard

Micky774 committed:
Merge branch 'dev' into zain/aiter-native-bshd-thd

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp:
Add a std::endl like in bwd debugging printing?

Micky774 replied:
Sorry not sure what you mean here, could you clarify?

wangye805 replied:
Like this: 
<img width="646" height="429" alt="image" src="https://github.com/user-attachments/assets/0db0de2c-2b1e-4f08-abcc-a34741468c8b" />

Just wanted to align the printing across fwd/bwd

Micky774 replied:
Done

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp:
```diff
@@ -977,26 +1030,44 @@ hipError_t ck_attn_varlen_bwd(
+  // modify the max_seqlen_q for better performance in 0-length cases
+  // lse_thd_ptr used as buffer
+  if(const char* env_p = std::getenv("NVTE_CK_RUNTIME_MAX_SEQLEN")) {
+    if(std::string(env_p) == "1" && !is_v3_api_check){
```
If we didn't set this env during runtime, will we get a nullptr for env_p? Is it safe to pass nullptr to std::string constructor? https://stackoverflow.com/questions/10771864/assign-a-nullptr-to-a-stdstring-is-safe

Micky774 replied:
The first `if` statement guards against that, so `env_p` will never be null in the string assignment

wangye805 replied:
Oh, I see. Got it

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -673,7 +685,8 @@ void fused_attn_ck_fwd_impl(
     std::cout<<"o_stride: ("<<o_stride[0]<<", "<<o_stride[1]<<", "<<o_stride[2]<<", "<<o_stride[3]<<"), ";
-    std::cout<<"pad_between_seqs: "<<pad_between_seqs<<", ";
+    std::cout<<"is_batch: "<<is_batch<<", ";
+    std::cout<<"is_padding: "<<is_padding<<", ";
```
Let's also print your newly added boolean flow variables like bshd_to_thd

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -691,11 +704,18 @@ void fused_attn_ck_fwd_impl(
   }
-  if(pad_between_seqs){
+  // If input is BSHD, we may directly convert to THD
+  if(bshd_to_thd){
```
You can move this generate_cu_seqlen_padded to after line 666. Therefore the workspace size calculation and workspace allocation branch conditions can align

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -705,26 +725,52 @@ void fused_attn_ck_fwd_impl(
+         devPtrV, 
+         v_stride[1], v_stride[2],
+         devPtrCuSeqlensQ, devPtrCuSeqlensKV, 
+         devPtrSeqOffsetsQ, devPtrSeqOffsetsKV, 
```
Well, I'm still against changing the input parameter values. Technically devPtrSeqOffsetsQ's should be read only.

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -1030,21 +1100,21 @@ void fused_attn_ck_bwd_impl(
-    workspace_next = static_cast<void *>(static_cast<int8_t *>(workspace_next) + h*max_tokens_q*sizeof(float));
+  }else if(bshd_to_thd){
+    // cu_seqlen_padded ptrs for THD conversion
+    devPtrSeqOffsetsQ = workspace_next;
```
Same here

wangye805 commented on transformer_engine/jax/csrc/extensions/attention.cpp:
```diff
@@ -235,18 +235,13 @@ pybind11::tuple GetFusedAttnForwardWorkspaceSizes(
   auto bias_shape = std::vector<size_t>{bias_batch, bias_heads, q_max_seqlen, kv_max_seqlen}; \
   size_t num_segments = input_batch;                                                          \
   if (is_ragged) {                                                                            \
-    auto cudnn_runtime_version = cudnnGetVersion();                                           \
```
We need a env to guard this runtime segment checking as well

wangye805 replied:
Thanks for adding this env. 

Is it possible to keep the original NV upstream behavior by adding another guard via #ifndef USE_ROCM? So that for NV code, they still check whether it's using runtime segment check when cudnn version >=9.3, and for us, we use a env

Micky774 committed:
Updated ptr variables and streamlined dispatch

Micky774 committed:
Added env guard

Micky774 committed:
Corrected bshd_to_thd conversion arguments

Micky774 committed:
Corrected logical flow

Micky774 committed:
Guarded memset and corrected allocation

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -671,11 +675,10 @@ void fused_attn_ck_fwd_impl(
-  NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrO, 0, o_storage_bytes, stream));
+  if(is_batch && is_padding){
+    // reset the final results since padded places need to be 0
+    NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrO, 0, o_storage_bytes, stream));
```
I suspect that devPtrO can skip the memset at all for current aiter/ck even in the padding mode. Let's have a try.

Micky774 replied:
As discussed, for now I have guarded the memsets and we shall follow up with another PR to update our pytests to ignore padding in batch mode.

wangye805 replied:
According to our discussion with Wen, for the fwd output O, we should add an env to let customer choose whether to memset the O or not

wangye805 reviewed (changes requested):
Generally, I think we can try to remove all memset except for dq, dq_acc. We can confirm with aiter/ck people

Micky774 committed:
Remove V3 API check and guard memsets

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp:
```diff
@@ -977,26 +1029,41 @@ hipError_t ck_attn_varlen_bwd(
-                                         is_v3_atomic_fp32,
-                                         how_v3_bf16_cvt);
-  if(average_runtime < 0){
+  float average_runtime_or_v3_check_status = aiter::mha_bwd(fmha_args,
```
As we are not checking the v3 api status and we are feeding in nullptr for cu_seqlen_padded ptrs, we can revert to our previous code (leaving cu_seqlen_ptr and v3 api check as default parameters)

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp:
```diff
@@ -1040,8 +1112,9 @@ hipError_t ck_attn_varlen_bwd(
           dk_or_dv_reduce_thd<CK_TILE_TYPE>, grid, block_dk, 0, stream,
-          h, hg, d_qk,
-          static_cast<const int32_t*>(cu_seqlen_kv_ptr)+b,
+          b, h, hg, d_qk,
```
Do we have gqa/mqa + MLA testcases w and w/ padding? If not, can we create those to verify this flow is actually working

Micky774 replied:
Will work on trying to add one in the JAX side -- for now I've added one on the TE side that isn't able to run due to too few backends supporting it, but that may change e.g. as we update AOTriton

wangye805 replied:
Thanks. Then let's skip the pytorch side gqa/mqa + MLA test for now. You can put a to-do here and add it later when other backends support it

Micky774 replied:
The test is already automatically skipped due to too few backends supporting it, so I'll leave it as-is. I've added a `TODO` comment for clarity though.

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp:
```diff
@@ -337,6 +371,8 @@ void log_bwd_config(const char* func_name,
     std::cout<<"uses_bwd_v3: "<<uses_bwd_v3<<std::endl;
     std::cout<<"is_v3_atomic_fp32: "<<is_v3_atomic_fp32<<std::endl;
     std::cout<<"how_v3_bf16_cvt: "<<how_v3_bf16_cvt<<std::endl;
+    std::cout<<"cu_seqlen_q_padded_ptr: "<<cu_seqlen_q_padded_ptr<<std::endl;
```
Similar to my comment in aiter::mha_fwd, we can skip printing cu_seqlen_padded ones if we always pass in nullptr's

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp:
```diff
@@ -633,7 +678,7 @@ hipError_t ck_attn_bwd(
 
   // print ck traits and args when needed
-  log_bwd_config(__FUNCTION__, data_type_str, is_group_mode, mask_type, bias_type, has_dbias, has_dropout, s_randval, deterministic, uses_bwd_v3, is_v3_atomic_fp32, how_v3_bf16_cvt, fmha_args);
+  log_bwd_config(__FUNCTION__, data_type_str, is_group_mode, mask_type, bias_type, has_dbias, has_dropout, s_randval, deterministic, uses_bwd_v3, is_v3_atomic_fp32, how_v3_bf16_cvt, nullptr, nullptr, fmha_args);
```
Revert back if we always pass in nullptrs to aiter::mha_fwd?

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_fwd.cpp:
```diff
@@ -271,7 +289,11 @@ hipError_t ck_attn_fwd(
-                                         uses_fwd_v3);
+                                         uses_fwd_v3, 
+                                         how_v3_bf16_cvt,
+                                         nullptr, //cu_seqlen_q_padded
```
Similar here

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_fwd.cpp:
```diff
@@ -258,7 +276,7 @@ hipError_t ck_attn_fwd(
   
   // print ck traits and args when needed
-  log_fwd_config(__FUNCTION__, data_type_str, is_group_mode, has_logits_soft_cap, mask_type, bias_type, has_lse, has_dropout, is_v_rowmajor, do_fp8_static_quant, uses_fwd_v3, fmha_args);
+  log_fwd_config(__FUNCTION__, data_type_str, is_group_mode, has_logits_soft_cap, mask_type, bias_type, has_lse, has_dropout, is_v_rowmajor, do_fp8_static_quant, uses_fwd_v3, how_v3_bf16_cvt, nullptr, nullptr, fmha_args);
```
Similar here

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_fwd.cpp:
```diff
@@ -431,23 +454,37 @@ hipError_t ck_attn_varlen_fwd(
+  }
   // print ck traits and args when needed
-  log_fwd_config(__FUNCTION__, data_type_str, is_group_mode, has_logits_soft_cap, mask_type, bias_type, has_lse, has_dropout, is_v_rowmajor, do_fp8_static_quant, uses_fwd_v3, fmha_args);
+  log_fwd_config(__FUNCTION__, data_type_str, is_group_mode, has_logits_soft_cap, mask_type, bias_type, has_lse, has_dropout, is_v_rowmajor, do_fp8_static_quant, uses_fwd_v3, how_v3_bf16_cvt, cu_seqlen_q_padded_ptr, cu_seqlen_kv_padded_ptr, fmha_args);
```
Similar here

wangye805 commented on transformer_engine/common/ck_fused_attn/src/ck_fused_attn_fwd.cpp:
```diff
@@ -431,23 +454,37 @@ hipError_t ck_attn_varlen_fwd(
-                                         has_lse,
-                                         uses_fwd_v3);
-  if(average_runtime < 0){
+  float average_runtime_or_v3_check_status = aiter::mha_fwd(
```
Similar here

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -621,10 +629,18 @@ void fused_attn_ck_fwd_impl(
-    // next h*max_tokens_q*sizeof(float) in workspace are for lse buffer
-    devPtrSoftmaxLSEWithoutPadding = workspace_next;
-    workspace_next = static_cast<void *>(static_cast<int8_t *>(workspace_next) + h*max_tokens_q*sizeof(float));
+  void* ptrSeqOffsetsQ = devPtrSeqOffsetsQ;
```
I think NV upstream use "dev" since those ptrs are on device side. Removing them could confuse code readers. How about name it as devPtrCuSeqlenPaddedQ/KV?

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -621,10 +629,18 @@ void fused_attn_ck_fwd_impl(
+  void* ptrSeqOffsetsQ = devPtrSeqOffsetsQ;
+  void* ptrSeqOffsetsKV = devPtrSeqOffsetsKV;
+
+
```
one extra empty line

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -873,12 +918,15 @@ void fused_attn_ck_bwd_impl(
     // just memset all dq, dk, dv
-    NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrdQ, 0, q_storage_bytes + k_storage_bytes+ v_storage_bytes, stream));
+    if(is_batch && is_padding){
+      NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrdQ, 0, q_storage_bytes + k_storage_bytes+ v_storage_bytes, stream));
```
Somehow I recall that ck/aiter requires dq memset for BSHD/SBHD without padding. That's why we have comment in line 915. Can you double check in our pytests? Perhaps we can reach out to CK/aiter team to confirm their nowadays behavior as well

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -966,10 +1014,12 @@ void fused_attn_ck_bwd_impl(
-  if(pad_between_seqs){
-    devPtrSoftmaxLSEWithoutPadding = workspace_next;
-    workspace_next = static_cast<void *>(static_cast<int8_t *>(workspace_next) + h*max_tokens_q*sizeof(float));
+  void* ptrSeqOffsetsQ = devPtrSeqOffsetsQ;
```
See the same naming suggestions above

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -873,12 +918,15 @@ void fused_attn_ck_bwd_impl(
-    if(pad_between_seqs){
+    if(is_batch && is_padding){
+      // HD_2HD, HD_H2D, HD_HD_HD can just memset dq itself
+      NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrdQ, 0, q_storage_bytes, stream));
```
I think the original logic is better:

memset dq for all cases
if (is_batch && is_padding){
  memset dk and dv as well
}
Unless ck/aiter changed their behavior

wangye805 replied:
After double checking, my previous words are not precise. Based on our discussion with Wen:
1). Without padding/unpadding workaround, dq requires memsetting since ck/aiter requires it. 
2). with padding/unpadding workaround (currently sbhd + padding), even if dq_without_padding will be memset to meet ck requirement, dq still requires memset according to Wen's figure

Micky774 replied:
Updated now to always be memset

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -1030,22 +1080,28 @@ void fused_attn_ck_bwd_impl(
+    ptrSeqOffsetsKV = workspace_next;
+    workspace_next = static_cast<void *>(static_cast<int8_t *>(workspace_next) + (b+1)*sizeof(int32_t));
+    generate_cu_seqlen_padded(s_q, s_kv, b, ptrSeqOffsetsQ, ptrSeqOffsetsKV, stream);
+    if(nvte_log_ck_config){
```
Move this debug message prinint above generate_cu_seqlen_padded and change it to more detailed: "\nattn_bwd(ck): generating cu_seqlen_padded in BSHD+padding to THD+padding"

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -1030,22 +1080,28 @@ void fused_attn_ck_bwd_impl(
-  bool nvte_ck_uses_bwd_v3 = getenv<int>("NVTE_CK_USES_BWD_V3", 0);
-  bool nvte_ck_is_v3_atomic_fp32 = getenv<int>("NVTE_CK_IS_V3_ATOMIC_FP32", 1);
-  int nvte_ck_how_v3_bf16_cvt = getenv<int>("NVTE_CK_HOW_V3_BF16_CVT", 1);
+
```
One extra line

wangye805 commented on transformer_engine/jax/csrc/extensions/attention.cpp:
```diff
@@ -234,15 +234,18 @@ pybind11::tuple GetFusedAttnForwardWorkspaceSizes(
   auto bias_shape = std::vector<size_t>{bias_batch, bias_heads, q_max_seqlen, kv_max_seqlen}; \
   size_t num_segments = input_batch;                                                          \
-  if (is_ragged) {                                                                            \
+  if(is_ragged){                                                                              \
```
Better keep the original format for this line to align with NV upstream

wangye805 commented on transformer_engine/jax/csrc/extensions/attention.cpp:
```diff
@@ -234,15 +234,18 @@ pybind11::tuple GetFusedAttnForwardWorkspaceSizes(
+    if(const char* env_p = std::getenv("NVTE_CK_RUNTIME_NUM_SEGMENTS")){                      \
+      use_runtime_num_segments_check = std::string(env_p) == "1";                             \
+    }                                                                                         \
+    if(cudnn_runtime_version < 90300 || use_runtime_num_segments_check){                      \
```
Emm, under this change, if people are running on NV upstream, they will be able to use our runtime segment feature. Originally I have the following logic in our mind:
if(cudnn_runtime_version >=90300){
  num_segments = input_batch*max_segments_per_seq;
  if(rocm_platform and env(NVTE_CK_RUNTIME_NUM_SEGMENTS)){
    run the runtime segment check
  }
}else{
  NV's old buggy runtime segment check
}

wangye805 replied:
But I guess it's fine since we never try to run our code on NV platform

Micky774 committed:
PR comments

wangye805 commented:
Let's also add how to use the runtime segment/max seqlen in readme under https://github.com/ROCm/TransformerEngine?tab=readme-ov-file#fused-attention-backends-on-rocm. Remind our customers that this will break the cudagraph

Micky774 committed:
Updated documentation

Micky774 commented:
> Let's also add how to use the runtime segment/max seqlen in readme under https://github.com/ROCm/TransformerEngine?tab=readme-ov-file#fused-attention-backends-on-rocm. Remind our customers that this will break the cudagraph

@wangye805 I've now updated the readme, but let me know if you have specific thoughts on it.

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
Let's try to remove this memsetting to see if ck/aiter still requires them nowadays

wangye805 reviewed (changes requested):
Please take a look at several unresolved conversation previously

Micky774 committed:
PR review reconciliation

- Updated debug message for BSHD-->THD conversion
- Added env variable to gate FWD output memset for padding
- Removed guards on memsets for d{Q,K,V} matrices

wenchenvincent commented:
@Micky774 Could you rebase/merge latest dev to incorporate the hot fixes for sgpu tests?

Micky774 committed:
Added explicit test

Micky774 committed:
Merge branch 'dev' into zain/aiter-native-bshd-thd

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -877,16 +923,13 @@ void fused_attn_ck_bwd_impl(
-        NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrdK, 0, k_storage_bytes, stream));
-        NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrdV, 0, v_storage_bytes, stream));
-      }
+    if(layout_group==NVTE_QKV_Layout_Group::NVTE_HD_2HD ||layout_group==NVTE_QKV_Layout_Group::NVTE_HD_H2D){
```
Do we still need memsetting dk/dv for cases without padding?

wangye805 commented:
pytorch test_numerics also shows some fused-attn related failures: 
FAILED tests/pytorch/test_numerics.py::test_kv_cache_accuracy[False-FusedAttention-TransformerLayer-sbhd-False-126m-1-dtype1] - AssertionError: Outputs not close enough in tensor at idx=0. Maximum difference at location [0, 650] with -0.90625 vs 0.5654296875 (diff 1.4716796875).

Not sure whether this is related to our decision to remove memsettings.

Micky774 committed:
Formatting for bwd debug

Micky774 committed:
Resolved error when using mixed formats e.g. sbhd_2bshd

Micky774 committed:
Updated guard on flash-attention forced support

Micky774 committed:
Added check for SBHD_2BSHD

Micky774 commented:
> pytorch test_numerics also shows some fused-attn related failures: FAILED tests/pytorch/test_numerics.py::test_kv_cache_accuracy[False-FusedAttention-TransformerLayer-sbhd-False-126m-1-dtype1] - AssertionError: Outputs not close enough in tensor at idx=0. Maximum difference at location [0, 650] with -0.90625 vs 0.5654296875 (diff 1.4716796875).
> 
> Not sure whether this is related to our decision to remove memsettings.

Those failures were due to a mix of not correctly dispatching to the `is_SBHD` workflow when dealing with `SBHD_2BSHD` formats, and miscalculating stride in the case of the same format. Resolved now.

Micky774 committed:
Added guard on dk/dv memset

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -705,25 +731,27 @@ void fused_attn_ck_fwd_impl(
         max_tokens_q,
         devPtrQWithoutPadding,
-        q_stride[1], (is_ragged? q_stride[2] : std::min(q_stride[0], q_stride[2])),
+        q_stride[1], q_stride[0],
```
Do we need std::min(q_stride[0], q_stride[2]) for SBHD_2BSHD formats with padding?

Micky774 replied:
No, we'd only need it for `BSHD_2SBHD` but we don't currently support/test that. If we want to support it, then I think it would best be done in a follow-up PR which generally refactors to account for heterogeneous q + k/v formats.

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -1074,50 +1127,50 @@ void fused_attn_ck_bwd_impl(
         max_tokens_q, max_tokens_kv,
         devPtrQWithoutPadding,
-        q_stride[1], (is_ragged? q_stride[2] : std::min(q_stride[0], q_stride[2])),
+        q_stride[1], q_stride[0],
```
Similar question here for SBHD_2BSHD "hybrid-style" format

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -877,8 +927,7 @@ void fused_attn_ck_bwd_impl(
     NVTE_CHECK_CUDA(cudaMemsetAsync(devPtrdQ, 0, q_storage_bytes, stream));
-    // for pad between seqs case, we need to reset all dq, dk, dv
-    if(pad_between_seqs){
+    if(is_padding && nvte_ck_zero_out_pad){
```
Based on Wen's analysis, dq, dk, dv requires zero out padding locations for subsequent grad computation. So they should be memset without conditions (nvte_ck_zero_out_pad can only control O tensor)

Micky774 replied:
Done

wangye805 reviewed (changes requested):
For those newly added hybrid qkv formats in upstream (NVTE_SBHD_2BSHD, NVTE_BSHD_2SBHD, NVTE_THD_2BSHD, and NVTE_THD_2SBHD): in addition to the SBHD_2BSHD pytest failures, are we able to correctly handle all other 3? Or is there only SBHD_2BSHD pytests now? 

NV upstream is separating format and is_ragged on q/kv and do subsequent processings accordingly: https://github.com/ROCm/TransformerEngine/blob/32e2d1d4d4af5976970dc3ca21bb17c97821a239/transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu#L79-L82

Maybe we can try similar technique. If I recall correctly, we need padding/unpadding for just q in SBHD_2BSHD and for just k/v in BSHD_2SBHD. 

Or it's okay if you want to leave this for another PR. 

By the way, there is an "extra line" comment you may have ignored :-)

Micky774 committed:
Removed env var gating for dk/dv zero padding, formatting

Micky774 committed:
Added inline comment to test

Micky774 committed:
Merge branch 'dev' into zain/aiter-native-bshd-thd

Micky774 committed:
Merge branch 'dev' into zain/aiter-native-bshd-thd

Micky774 committed:
Corrected Softmax LSE buffer allocation

wangye805 commented on transformer_engine/common/fused_attn_rocm/fused_attn_ck.cpp:
```diff
@@ -817,10 +867,8 @@ void fused_attn_ck_bwd_impl(
 
   // Exit to request upper level API to allocate memory if needed
   if(workspace==nullptr){
-    size_t workspace_size_lse = max_tokens_q*h*sizeof(float);
```
I just noticed this line removal. This is the lse workspace for all configs: https://github.com/ROCm/TransformerEngine/blob/0018b1bce0240ad53a6ac24a3b8052d9f067d34b/transformer_engine/common/ck_fused_attn/src/ck_fused_attn_bwd.cpp#L568. It's different from the softmax_lse buffer that used for padding/unpadding (conversion between ck/aiter and TE api)

Micky774 committed:
Correct Softmax LSE buffer memory allocation

Micky774 committed:
Adjusted fwd pass softmax lse allocation

Micky774 committed:
Adjusted bwd pass softmax conversion allocation

Micky774 committed:
Minor reversions

wangye805 commented:
In fact, I saw some level 3 pytorch cp pytest failures by run level 3 ci locally: 
```
=========================== short test summary info ============================
FAILED tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[False-p2p-thd-cp_1_0-bf16] - subprocess.CalledProcessError: Command '['python3', '-m', 'torch.distributed.launch', '--nproc-per-node=2', '/workspace/te_native_bshd_thd/tests/pytorch/fused_attn/run_fused_attn_with_cp.py', 'dtype=bf16', 'model=cp_1_0', 'qkv_format=thd', 'kernel_backend=FusedAttention', 'cp_comm_type=p2p', 'fp8_mha=False']' returned non-zero exit status 1.
FAILED tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[False-p2p-thd-cp_1_1-bf16] - subprocess.CalledProcessError: Command '['python3', '-m', 'torch.distributed.launch', '--nproc-per-node=2', '/workspace/te_native_bshd_thd/tests/pytorch/fused_attn/run_fused_attn_with_cp.py', 'dtype=bf16', 'model=cp_1_1', 'qkv_format=thd', 'kernel_backend=FusedAttention', 'cp_comm_type=p2p', 'fp8_mha=False']' returned non-zero exit status 1.
FAILED tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[False-p2p-thd-cp_2_0-bf16] - subprocess.CalledProcessError: Command '['python3', '-m', 'torch.distributed.launch', '--nproc-per-node=2', '/workspace/te_native_bshd_thd/tests/pytorch/fused_attn/run_fused_attn_with_cp.py', 'dtype=bf16', 'model=cp_2_0', 'qkv_format=thd', 'kernel_backend=FusedAttention', 'cp_comm_type=p2p', 'fp8_mha=False']' returned non-zero exit status 1.
FAILED tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[False-p2p-thd-cp_2_1-bf16] - subprocess.CalledProcessError: Command '['python3', '-m', 'torch.distributed.launch', '--nproc-per-node=2', '/workspace/te_native_bshd_thd/tests/pytorch/fused_attn/run_fused_attn_with_cp.py', 'dtype=bf16', 'model=cp_2_1', 'qkv_format=thd', 'kernel_backend=FusedAttention', 'cp_comm_type=p2p', 'fp8_mha=False']' returned non-zero exit status 1.
SKIPPED [48] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:68: CP implementation with KV P2P does not support sliding window yet!
SKIPPED [16] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:70: CP implementation with KV all-gather does not support THD format yet!
SKIPPED [24] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:74: CP implementation with QKVO A2A does not support THD format yet!
SKIPPED [240] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:133: FP8 attention has not been supported on ROCm yet!
SKIPPED [40] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:153: CP implementation with KV P2P does not support sliding window yet!
SKIPPED [64] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:137: THD format does not support post_scale_bias yet!
SKIPPED [32] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:155: CP implementation with KV all-gather does not support bias yet!
SKIPPED [24] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:139: CP implementation with KV all-gather does not support THD format yet!
SKIPPED [64] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:157: CP implementation with QKVO A2A does not support bias yet!
SKIPPED [48] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:141: CP implementation with QKVO A2A does not support THD format yet!
SKIPPED [104] tests/pytorch/fused_attn/test_fused_attn_with_cp.py:164: Only fp8 works with fp8_mha=True!
===== 4 failed, 204 passed, 704 skipped, 2 warnings in 3065.08s (0:51:05) ======
Error in test [ck] fused_attn/test_fused_attn_with_cp.py
Done [ck] fused_attn/test_fused_attn_with_cp.py
Got 1 test errors during run at level 3
```
Attached you can find the detailed log
[torch_mgpu.txt](https://github.com/user-attachments/files/23924167/torch_mgpu.txt)

wangye805 committed:
[ROCm] fix the aiter fwd v3 cu_seqlen/cu_seqlen_padded api issue

wangye805 committed:
Update README.rst to fix formatting

Micky774 committed:
Merge branch 'dev' into zain/aiter-native-bshd-thd

wangye805 committed:
Merge remote-tracking branch 'origin/dev' into zain/aiter-native-bshd-thd

wangye805 committed:
[ROCm] update aiter commit with swa fix

wangye805 committed:
Merge remote-tracking branch 'origin/dev' into zain/aiter-native-bshd-thd

wangye805 commented:
Manual CI runs in local MI300X machine:
[core_sgpu_l3.txt](https://github.com/user-attachments/files/24594689/core_sgpu_l3.txt)
[torch_mgpu_l3.txt](https://github.com/user-attachments/files/24594703/torch_mgpu_l3.txt)
[jax_sgpu_l3.txt](https://github.com/user-attachments/files/24594698/jax_sgpu_l3.txt)

wangye805 commented:
[jax_mgpu_l3.txt](https://github.com/user-attachments/files/24594739/jax_mgpu_l3.txt)
